src/Compilers/CSharp/Portable/Lowering/Instrumentation/ModuleCancellationInstrumenter.cs

```diff
@@ -0,0 +1,224 @@
+            method = cancellableOverload;
+            arguments = [.. arguments, _factory.ModuleCancellationToken()];
+            argumentRefKindsOpt = argumentRefKindsOpt.IsDefault ? default : [.. argumentRefKindsOpt, RefKind.None];
+        }
+    }
+
+    /// <summary>
+    /// Find an overload whose last parameter is <see cref="CancellationToken"/> and
+    /// the parameter types and ref kinds of all other parameters match those of <paramref name="method"/>.
+    /// </summary>
+    private MethodSymbol? FindOverloadWithCancellationToken(MethodSymbol method)
+    {
+        // Switching to a cancellable overload only applies to ordinary methods and constructors.
+        if (method.MethodKind is not (MethodKind.Ordinary or MethodKind.Constructor))
+        {
+            return null;
+        }
+
+        // Look for an overload of the definition with no type parameter substitutions.
+        var methodDefinition = method.OriginalDefinition;
+
+        var methodsSetsRequiredMembers = methodDefinition.HasSetsRequiredMembers;
+
+        foreach (var member in methodDefinition.ContainingType.GetMembers(method.Name))
+        {
+            // If the member is the method being compiled, calling it could result in a recursion not present in the code previously.
+            // We can't guarantee the instrumentation won't result in infinite recursion, but we can avoid the trivial case.
+            if (member == _factory.TopLevelMethod?.OriginalDefinition)
+            {
+                continue;
+            }
+
+            // It's unlikely that real-world APIs have overloads that differ in dynamic,
+            // but if they do avoid selecting them since their intended use might differ.
+            //
+            // Similarly, we also only consider overloads with the same visibility.
+            // We could potentially allow the visibility to differ,
+            // but we'd need to check if the overload is accessible at the call site.
+
+            const TypeCompareKind TypeComparisonKind = TypeCompareKind.CLRSignatureCompareOptions & ~TypeCompareKind.IgnoreDynamic;
+
+            if (member.IsStatic == methodDefinition.IsStatic &&
+                member.MetadataVisibility == methodDefinition.MetadataVisibility &&
+                member is MethodSymbol { Parameters: [.., { RefKind: RefKind.None, Type: { } lastParamType }] parametersWithCancellationToken } overload &&
+                overload.Arity == methodDefinition.Arity &&
+                methodDefinition.Parameters.Length == parametersWithCancellationToken.Length - 1 &&
+                lastParamType.Equals(_throwMethod.ContainingType, TypeCompareKind.ConsiderEverything) &&
+                (!methodsSetsRequiredMembers || overload.HasSetsRequiredMembers))
+            {
+                // we are only considering constructors and ordinary methods, which have different name.
+                Debug.Assert(methodDefinition.MethodKind == overload.MethodKind);
```

Comment:
>methodDefinition.MethodKind == overload.MethodKind [](http://example.com/codeflow?start=29&length=50)

Since the kind comparison for ```overload``` got removed, I think this assumption is incorrect. For example, it looks like for the example below an invocation of `get_Item` method will be replaced with an invocation of an indexer accessor.

```
public class Test
{
    public int get_Item() => 0;
    
    public int this[System.Threading.CancellationToken t] => 0;
}
```

Reply:
Ah, good catch. I'll add the check back and also a test.